crates/red_knot/src/program/check.rs

```diff
@@ -187,87 +185,61 @@ impl CheckFileTask {
 
 #[derive(Clone, Debug)]
 struct CheckContext {
-    cancellation_token: CancellationToken,
-    sender: crossbeam_channel::Sender<CheckFileMessage>,
+    sender: crossbeam::channel::Sender<CheckFileMessage>,
 }
 
 impl CheckContext {
-    fn new(
-        cancellation_token: CancellationToken,
-        sender: crossbeam_channel::Sender<CheckFileMessage>,
-    ) -> Self {
-        Self {
-            cancellation_token,
-            sender,
-        }
+    fn new(sender: crossbeam::channel::Sender<CheckFileMessage>) -> Self {
+        Self { sender }
     }
 
     /// Queues a new file for checking using the [`CheckScheduler`].
     #[allow(unused)]
     fn schedule_check_file(&self, file_id: FileId) {
         self.sender.send(CheckFileMessage::Queue(file_id)).unwrap();
     }
-
-    /// Returns `true` if the check has been cancelled.
-    fn is_cancelled(&self) -> bool {
-        self.cancellation_token.is_cancelled()
-    }
-
-    fn cancelled_ok(&self) -> Result<(), CheckError> {
-        if self.is_cancelled() {
-            Err(CheckError::Cancelled)
-        } else {
-            Ok(())
-        }
-    }
 }
 
 struct CheckFilesLoop<'a> {
     scheduler: &'a dyn CheckScheduler,
-    cancellation_token: CancellationToken,
     pending: usize,
     queued_files: FxHashSet<FileId>,
 }
 
 impl<'a> CheckFilesLoop<'a> {
-    fn new(scheduler: &'a dyn CheckScheduler, cancellation_token: CancellationToken) -> Self {
+    fn new(scheduler: &'a dyn CheckScheduler) -> Self {
         Self {
             scheduler,
-            cancellation_token,
-
             queued_files: FxHashSet::default(),
             pending: 0,
         }
     }
 
-    fn run(mut self, files: impl Iterator<Item = FileId>) -> Result<Vec<String>, CheckError> {
+    fn run(mut self, files: impl Iterator<Item = FileId>) -> QueryResult<Vec<String>> {
         let (sender, receiver) = if let Some(max_concurrency) = self.scheduler.max_concurrency() {
-            crossbeam_channel::bounded(max_concurrency.get())
+            crossbeam::channel::bounded(max_concurrency.get())
         } else {
             // The checks run on the current thread. That means it is necessary to store all messages
             // or we risk deadlocking when the main loop never gets a chance to read the messages.
-            crossbeam_channel::unbounded()
+            crossbeam::channel::unbounded()
         };
 
-        let context = CheckContext::new(self.cancellation_token.clone(), sender.clone());
+        let context = CheckContext::new(sender.clone());
 
         for file in files {
-            self.queue_file(file, context.clone())?;
+            self.queue_file(file, context.clone());
```

Comment:
This can deadlock when `files.len() > max_concurrency)`. I have a follow up PR to fix this.